src/modm/ui/color/rgb.hpp

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright (c) 2009, Martin Rosekeit
+ * Copyright (c) 2009-2013, Fabian Greif
+ * Copyright (c) 2012-2013, 2015, Niklas Hauser
+ * Copyright (c) 2013, David Hebbeker
+ * Copyright (c) 2021, Thomas Sommer
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_COLOR_RGB_HPP
+#define MODM_COLOR_RGB_HPP
+
+#include <stdint.h>
+
+#include <concepts>
+#include <modm/io/iostream.hpp>
+#include <modm/math/saturated/saturated.hpp>
+#include <modm/math/utils/arithmetic_traits.hpp>
+#include <modm/math/utils/concepts.hpp>
+#include <type_traits>
+
+#include "brightness.hpp"
+#include "hsv.hpp"
+#include "rgb565.hpp"
+
+using namespace std;
+namespace modm
+{
+namespace color
+{
+
+// forward declarations for convertion constructors
+template<std::unsigned_integral T>
+class HsvT;
+
+template<std::unsigned_integral T>
+class BrightnessT;
+
+class Rgb565;
+
+/**
+ * Color in HSV Colorspace
+ *
+ * @author	Martin Rosekeit, Fabian Greif, Niklas Hauser, David Hebbeker, Thomas Sommer
+ * @ingroup	modm_ui_color
+ */
+
+template<std::unsigned_integral T = uint8_t>
+class RgbT
+{
+public:
+	T red{0};
+	T green{0};
+	T blue{0};
+
+	using TSum = modm::WideType<T>;
+
+	constexpr RgbT() = default;
+
+	constexpr RgbT(T red, T green, T blue) : red(red), green(green), blue(blue) {}
+
+	/**
+	 * Convertion Constructor for HSV Color
+	 *
+	 * \param hsv	HSV Color
+	 */
+	template<std::unsigned_integral _T>
+	constexpr RgbT(const HsvT<_T>& hsv);
+
+	/**
+	 * Convertion Constructor for Brightness
+	 *
+	 * \param brightness	Brightness 'Color'-object
+	 */
+	// TODO Plump conversion, implement the right way
+	template<std::unsigned_integral _T>
+	constexpr RgbT(const BrightnessT<_T> brightness)
+		: red(brightness), green(brightness), blue(brightness)
+	{}
+
+	/**
+	 * Convertion Constructor for RGB565 Color
+	 *
+	 * \param rgb565	RGB565 Color
+	 */
+	constexpr RgbT(const Rgb565& rgb565)
+		: red((rgb565.color >> 8) & 0xF8),
+		  green((rgb565.color >> 3) & 0xFC),
+		  blue(rgb565.color << 3)
+	{}
+
+	inline TSum
+	getColorSum() const
+	{
+		return red + green + blue;
+	}
+
+	template<modm::fundamental IntermediateType = float, unsigned int multiplier = 100,
+			 std::unsigned_integral ReturnType = T>
+	ReturnType
+	getRelative(const T color) const
+	{
+		return IntermediateType(color) * IntermediateType(multiplier) /
+			   IntermediateType(getColorSum());
+	}
```

Comment:
Can you please give me an application example for relative colors, so i can add more head in future considerations.

Reply:
Good question… 

This was implemented for the Eurobot competition by @dhebbeker to distinguish between red and blue in bad lighting conditions. I'm not sure if it is particularly useful and would perhaps even just remove it from the class…
I feel like this would be better as an algorithm function that takes a RGB value and outputs a different RGB.